fpga_interchange/arch_pack_clusters.cc

```diff
@@ -370,6 +506,344 @@ static bool check_cluster_cells_compatibility(CellInfo *old_cell, CellInfo *new_
+                break;
+            for (const auto &arc : idx_to_cells){
+                if (arc.second.size() == 1)
+                    for (const auto &idx : possible_idx[*arc.second.begin()])
+                        if (idx != arc.first)
+                            removequeue.push(std::pair<uint32_t, CellInfo*>(idx, *arc.second.begin()));
+            }
+            while(!removequeue.empty()){
+                auto t = removequeue.front(); removequeue.pop();
+                uint32_t idx = t.first;
+                CellInfo *cell = t.second;
+                idx_to_cells[idx].erase(cell);
+                change = true;
+                changed_idxs.insert(idx);
+            }
+            for (const uint32_t &idx : changed_idxs)
+                for (const auto &connection : cluster->connection_graph[idx].connections)
+                    workqueue.push(std::pair<uint32_t, uint32_t>(idx, connection.target_idx));
+
+            binary_constraint_check(cluster, workqueue, idx_to_cells, ctx);
+        }while(change);
+        if(failed){
+            if(ctx->debug)
+                log_info("Single cell mapping failed\n");
+            continue;
+        }
+        if (ctx->debug){
+            log_info("After mapping indices with single cell\n");
+            dict<CellInfo *, pool<uint32_t>, hash_ptr_ops> possible_idx;
+            for (const auto &arc : idx_to_cells)
+                for (const auto &cell : arc.second)
+                    possible_idx[cell].insert(arc.first);
+
+            for (const auto arc : possible_idx){
+                log_info("Possible idx %s:\n", arc.first->name.c_str(ctx));
+                for (const auto idx : arc.second)
+                    log_info("    - %d\n", idx);
+            }
+        }
+        // At this point all indices that cloud only be mapped to single cell are mapped
+        // Next step is to run solver with backtracing to solve for other idx<->cell mappings
+        if (ctx->debug)
+            log_info("Back solver\n");
+        if(!back_solver(cluster, idx_to_cells, ctx)){
+            if(ctx->debug)
+                log_info("Back solver failed\n");
+            continue;
+        }
+        if (ctx->debug){
+            log_info("Final mapping after back solver\n");
+            dict<CellInfo *, pool<uint32_t>, hash_ptr_ops> possible_idx;
+            for (const auto &arc : idx_to_cells)
+                for (const auto &cell : arc.second)
+                    possible_idx[cell].insert(arc.first);
+
+            for (const auto arc : possible_idx){
+                log_info("Possible idx %s:\n", arc.first->name.c_str(ctx));
+                for (const auto idx : arc.second)
+                    log_info("    - %d\n", idx);
+            }
+        }
+        Cluster cluster_info;
+        cluster_info.root = ci;
+        cluster_info.index = index;
+        cluster_info.cluster_nodes.resize(idx_to_cells.size());
+        ci->cluster.set(ctx, ci->name.str(ctx));
```

Comment:
`ci->cluster = ci->name` should be fine here?

Reply:
Fixed in new commit